hubbard/hamiltonian.py

```diff
@@ -29,35 +28,48 @@ class HubbardHamiltonian(object):
     DM: `sisl.physics.DensityMatrix` instance, optional
         A spin-polarized density datrix generated with sisl
         to use as a initial spin-densities
-    U: float, optional
-        on-site Coulomb repulsion
+    U: float, np.ndarray optional
+        Coulomb repulsion parameter
     q: array_like, optional
         Two values specifying up, down electron occupations
     nkpt: array_like or `sisl.physics.BrillouinZone` instance, optional
         Number of k-points along (a1, a2, a3) for Monkhorst-Pack BZ sampling
     kT: float, optional
         Temperature of the system in units of the Boltzmann constant
+
+    Note
+    ----
+    The implementation to solve the extended Hubbard model (with inter-atomic interactions) has not been tested
     """
 
-    def __init__(self, TBHam, DM=0, U=0.0, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
+    def __init__(self, TBHam, DM=0, U=None, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
         """ Initialize HubbardHamiltonian """
 
         if not TBHam.spin.is_polarized:
             raise ValueError(self.__class__.__name__ + ' requires a spin-polarized system')
 
-        # Use sum of all matrix elements as a basis for hash function calls
         H0 = TBHam.copy()
         H0.shift(np.pi) # Apply a shift to incorporate effect of S
-        self.hash_base = H0.H.tocsr().sum()
-
-        self.U = U # hubbard onsite Coulomb parameter
+        self._hash_base = hash((H0.H.tocsr(0).data.tostring(), H0.H.tocsr(1).data.tostring()))
 
         # Copy TB Hamiltonian to store the converged one in a different variable
         self.TBHam = TBHam
         self.H = TBHam.copy()
         self.H.finalize()
         self.geometry = TBHam.geometry
 
+        if U is None:
+            try:
+                # Try to extract U stored in sisl.Geometry object
+                U = np.array(len(self.geometry))
+                for ia in self.geometry:
+                    U[ia] = self.geometry.atoms[ia].U
```

Comment:
perhaps this should be changed to
```python
# Try to extract U stored in sisl.Geometry object
￼                U = np.zeros(len(self.geometry))
￼                for ia in self.geometry:
try:
￼                    U[ia] = self.geometry.atoms[ia].U
except:
pass
￼```
Perhaps this is fine for now, but later it may be nice to change this to orbitals to allow multiple orbitals per atom, and thus have different U per orbital.